tests/utils.py

```diff
@@ -154,69 +161,170 @@ class TestServerProcess():
+        stdout=subprocess.PIPE, stderr=subprocess.STDOUT, cwd=popen_cwd)
 
-    self.__logger.info('Server process with process id ' \
-        + str(self.__server_process.pid) + " serving on port " \
-        + str(self.port) + ' started.')
 
-    if timeout > 0:
-      try:
-        wait_for_server('localhost', self.server, self.port, timeout)
-      except Exception as e:
-        # Make sure that errors from the server side will be logged.
+
+  def _start_redirect_thread(self):
+    """Starts a thread responsible to redirect stdout/stderr to the Queue."""
+
+    # Run log_queue_worker() in a thread.
+    # The thread will exit when the child process dies.
+    self._log_queue = queue.Queue()
+    log_thread = threading.Thread(target=self._log_queue_worker,
+        args=(self.__server_process.stdout, self._log_queue))
+
+    # "daemon = True" means the thread won't interfere with the process exit.
+    log_thread.daemon = True
+    log_thread.start()
+
+
+
+  def _log_queue_worker(self, stream, line_queue):
+    """
+    Worker function to run in a seprate thread.
+    Reads from 'stream' (stdout), puts lines in Queue (Queue is thread-safe).
+    """
+
+    while True:
+      # readline() is a blocking operation;
+      # We need to decode because readline returns a sequence of 8-bit bytes,
+      # but we want to put the messages in the queue as strings.
+      log_line = stream.readline().decode()
+      line_queue.put(log_line)
+
+      if len(log_line) == 0:
+        # This mean we are at the end of the stream meaning the server process
+        # has exited and we don't need the stream anymore.
+        stream.close()
+        break
+
+
+
+  def _set_port_if_in_logs(self, timeout):
+    """
+    Checks if the port is logged from the server subprocess.
+    If it's found, self.port is set and returns True.
+    Otherwise, returns False.
+    TestServerProcessError is throwned if the process has exited
+    or if it has failed to bind.
+    """
+
+    # We have hardcoded the message we expect on a successful server startup.
+    # This message should be the first message sent by the server!
+    expected_msg = 'bind succeeded, server port is: '
+    # Get lines until port is found, child process exits, or we timeout.
+    try:
+      line = self._log_queue.get(timeout=timeout)
+      if len(line) == 0:
+        # The process has exited.
+        raise TestServerProcessError(self.server + ' exited unexpectedly ' \
+            + 'with code ' + str(self.__server_process.poll()) + '!')
+
+      elif line.startswith(expected_msg):
+        self.port = int(line[len(expected_msg):])
+        return True
+      else:
+        # An exception or some other message is printed from the server.
+        self.__logged_messages.append(line)
+        # Check if more lines are logged.
         self.flush_log()
-        raise e
+        raise TestServerProcessError(self.server + ' failed to bind!')
```

Comment:
```suggestion
        raise TestServerProcessError(self.server + ' did not print port message as first stdout line as expected')
```
this might be fairer for the future testserver writer whose server binds just fine but does not print what we expect.